src/atlas_assistant/tools/sql.py

```diff
@@ -5,42 +5,99 @@
 from pydantic import BaseModel
 
 from ..context import Context
+from ..dataset import Dataset
 from ..state import State
 
 
-class SqlQuery(BaseModel):
-    query: str
-    """The SQL query"""
+class SqlQueryParts(BaseModel):
+    select: str
+    """The SQL select statement"""
+
+    where: str
+    """The SQL where statement"""
+
+    group_by: str | None
+    """The SQL group by statement"""
+
+    order_by: str | None
+    """The SQL order by statement"""
+
+    limit: str | None
+    """The SQL limit statement"""
 
     explanation: str
     """An explanation of why the model generated this query"""
 
+    def get_query(self, href: str) -> str:
+        """Gets the full SQL query"""
+        parts = [f"SELECT {self.select} FROM '{href}' WHERE {self.where}"]
+        if self.group_by:
+            parts.append(f"GROUP BY {self.group_by}")
+        if self.order_by:
+            parts.append(f"ORDER BY {self.order_by}")
+        if self.limit:
+            parts.append(f"LIMIT {self.limit}")
+        return " ".join(parts)
+
 
 @tool
 def generate_sql(query: str, runtime: ToolRuntime[Context, State]) -> Command[None]:
     """Generates SQL to query a dataset.
 
+    There must be a selected dataset to generate SQL.
+
     Args:
         query: The question that we're going to answer with an SQL query.
     """
+    dataset = runtime.state["dataset"]
+    if not dataset:
+        return Command(
+            update={
+                "messages": [
+                    ToolMessage(
+                        content="No dataset selected", tool_call_id=runtime.tool_call_id
+                    )
+                ]
+            }
+        )
+    s3_href = dataset.s3_href
+    if not s3_href:
+        return Command(
+            update={
+                "dataset": None,
+                "messages": [
+                    ToolMessage(
+                        content=f"Dataset {dataset.item.id} does not have an s3 href, "
+                        "choose a different dataset",
+                        tool_call_id=runtime.tool_call_id,
+                    )
+                ],
+            }
+        )
+
     settings = runtime.context.settings
     assert settings.mistral_api_key
-    client = Mistral(api_key=settings.mistral_api_key.get_secret_value())
+    client = Mistral(
+        api_key=settings.mistral_api_key.get_secret_value()
+        if settings.mistral_api_key
+        else None
+    )
     response = client.chat.parse(
         model="codestral-latest",
         messages=[
             {
                 "role": "system",
-                "content": get_prompt(query),
+                "content": get_prompt(dataset),
```

Comment:
If we could inject dataset specific instructions that could help too.

For instance, in files that have scenarios, we probably always need to either select a scenario or group by them to avoid summing over multiple scenarios at the same time.